File: app/src/main/java/com/thewizrd/simpleweather/widgets/WeatherWidgetProvider4x2Clock.java
package com.thewizrd.simpleweather.widgets;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;

import com.thewizrd.simpleweather.App;
import com.thewizrd.simpleweather.R;

import static com.thewizrd.simpleweather.widgets.WidgetType.Widget4x2Clock;

public class WeatherWidgetProvider4x2Clock extends WeatherWidgetProvider {
    private static WeatherWidgetProvider4x2Clock sInstance;

    public static synchronized WeatherWidgetProvider4x2Clock getInstance() {
        if (sInstance == null)
            sInstance = new WeatherWidgetProvider4x2Clock();

        return sInstance;
    }

    // Overrides
    @Override
    public WidgetType getWidgetType() {
        return Widget4x2Clock;
    }

    @Override
    public int getWidgetLayoutId() {
        return R.layout.app_widget_4x2_clock;
    }

    @Override
    protected String getClassName() {
        return this.getClass().getName();
    }

    @Override
    public ComponentName getComponentName() {
        return new ComponentName(App.getInstance().getAppContext(), getClassName());
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        String action = intent == null ? null : intent.getAction();

        if (Intent.ACTION_TIME_CHANGED.equals(action)
                || Intent.ACTION_TIMEZONE_CHANGED.equals(action)) {
            // Update clock widget
            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            ComponentName componentname = new ComponentName(context.getPackageName(), getClassName());
            int[] appWidgetIds = appWidgetManager.getAppWidgetIds(componentname);

            refreshClock(context, appWidgetIds);
            refreshDate(context, appWidgetIds);
        } else {
            super.onReceive(context, intent);
        }
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/widgets/WeatherWidgetProvider4x2Clock.kt
package com.thewizrd.simpleweather.widgets

import com.thewizrd.simpleweather.R

class WeatherWidgetProvider4x2Clock : WeatherWidgetProvider() {
    override val info: WidgetProviderInfo by lazy { Info.getInstance() }

    class Info private constructor() : WidgetProviderInfo() {
        override val className: String
            get() = WeatherWidgetProvider4x2Clock::class.java.name
        override val widgetType: WidgetType
            get() = WidgetType.Widget4x2Clock
        override val widgetLayoutId: Int
            get() = R.layout.app_widget_4x2_clock

        companion object {
            private var _instance: WidgetProviderInfo? = null

            @JvmStatic
            fun getInstance(): WidgetProviderInfo {
                if (_instance == null) {
                    _instance = Info()
                }
                return _instance!!
            }
        }
    }
}